File: Java/Hexdump.java
import java.io.IOException;

public class Hexdump {
    private static int col = 16;

    public static void main(String[] args) throws IOException {
        byte[] buf = new byte[col];
        int numRead;
        while ((numRead = System.in.read(buf)) > 0) {
            int i;
            for (i = 0; i < numRead; i++) {
                System.out.print(String.format("%02X ", buf[i]));
            }
            for (i = 0; i < (col - numRead) * 3 + 2; i++) {
                System.out.print(' ');
            }
            for (i = 0; i < numRead; i++) {
                if (buf[i] < 0x20 || buf[i] > 0x7E) {
                    System.out.print('.');
                } else {
                    System.out.print((char)buf[i]);
                }
            }
            System.out.println();
        }
    }
}

File: C/hexdump.c
#include <stdio.h>
#include <unistd.h>

const int cols = 16;

int main() {
    u_int8_t buf[cols];
    _ssize_t num_read;
    while ((num_read = read(STDIN_FILENO, buf, cols)) > 0) {
        int i;
        for (i = 0; i < num_read; i++) {
            printf("%02X ", buf[i]);
        }
        printf("%*c", (cols - num_read) * 3 + 2, ' ');
        for (i = 0; i < num_read; i++) {
            if (buf[i] < 0x20 || buf[i] > 0x7E) {
                printf(".");
            } else {
                printf("%c", (char)buf[i]);
            }
        }
        printf("\n");
    }
}